certora/specs/Delegation.spec

```diff
@@ -0,0 +1,61 @@
+import "ERC20.spec";
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sload uint256 votingPower (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] {
+    require sumOfVotingPower >= to_mathint(votingPower);
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] uint256 newValue (uint256 oldValue) {
+    sumOfVotingPower = sumOfVotingPower - oldValue + newValue;
+}
+
+// Check that zero address has no voting power assuming that zero address can't make transactions.
+invariant zeroAddressNoVotingPower()
+    delegatee(0x0) == 0x0 && delegatedVotingPower(0x0) == 0
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    { preserved with (env e) { require e.msg.sender != 0; } }
+
+// Check that the voting power is never greater than the total supply of tokens.
+invariant totalSupplyLTEqSumOfVotingPower()
```

Comment:
```suggestion
invariant totalSupplyEqualSumOfVotingPowerPlusZeroVotingPower()
```

Reply:
I think that the very long name is more accurate but less informative. The invariant we really care about is the the sum of the voting is less than or equal to the total supply of tokens, how we do we do this by counting _theoretical_ voting power of zero. In my view, the name should inform on the property being checked rather than how. Although a comment explaining how this is a achieved may be missing indeed.

Reply:
I agree that the most important is to have the property "total supply is greater than the sum of voting power". What about doing just that: rename the invariant as I suggested, and create another rule with a better name that uses this invariant to prove that property

Reply:
I fail to see how this approach improves the currents state of things.

Reply:
The goal is to not have an invariant whose name doesn't reflect what it is proving (current state of things). Also we want to make sure that the property we want is actually verified, so let's write that property verbatim: `assert totalSupply()) >= sumOfVotingPower`

Reply:
> In my view, the name should reflect the property being checked rather than how it is checked.

I think the name should reflect the property written: this is because we can later do `requireInvariant invariantName`, and it would be very confusing if `invariantName` is referencing something else than the formula being `require`d

Reply:
Changed in https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/f80c7fbaebcbaf0f9b92f71aeafeed59a04c15fc.